### PR TITLE
[tflite] make `//tensorflow/lite/kernels:all` build

### PR DESCRIPTION
**Pull request by freedomtan**
make
```
bazel build --config android_arm64 //tensorflow/lite/kernels:all
```
work.

I want to run all test cases on an Android devices, but  there are some problems why trying to build `//tensorflow/lite/kernels:all`. All of them are libm related problems.

### CODE REVIEW

**Comment by multiverse-tf** [comment]
Thx for reporting the issue! Could you paste the linking error output here? It's likely there's a build toolchain issue here that we didn't catch in our continuous integration tests.

**Comment by multiverse-tf** [comment]
@yyoon and @terryheo, just wondering if this libm linking issue with bazel rings any bell to you? Many thanks!

**Comment by freedomtan** [comment]
I'll check my ANDROID SDK and NDK settings.

FYR. For `acceleration_test_util_internal_test`, I got
```
ERROR: /hack/freedom/tensorflow/tf-clean/tensorflow/lite/kernels/BUILD:155:1: Linking of rule '//tensorflow/lite/kernels:acceleration_test_util_internal_test' failed (Exit 1)
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Sstrings_Slibstrings.so: undefined reference to `nan'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Sbase_Slibexponential_Ubiased.so: undefined reference to `log2'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Stime_Slibtime.so: undefined reference to `modf'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Sstrings_Slibstrings.so: undefined reference to `frexp'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Sstrings_Slibstrings.so: undefined reference to `nanf'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Snumeric_Slibint128.so: undefined reference to `truncl'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Snumeric_Slibint128.so: undefined reference to `ldexpl'
bazel-out/arm64-v8a-opt/bin/_solib_arm64-v8a/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Sstrings_Slibstrings.so: undefined reference to `ldexpf'
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

**Comment by freedomtan** [comment]
my ndk and sdk settings `.tf_configure.bazelrc`
```
uild --action_env ANDROID_NDK_HOME="/home/freedom/work/android-ndk-r20"
build --action_env ANDROID_NDK_API_LEVEL="28"
build --action_env ANDROID_BUILD_TOOLS_VERSION="29.0.2"
build --action_env ANDROID_SDK_API_LEVEL="28"
build --action_env ANDROID_SDK_HOME="/home/freedom/work/android-sdk-linux"
```

**Comment by yyoon** [comment]
@freedomtan 

I tend to think that it's probably better to add the linker flags to the final binary/test targets rather than the intermediate library targets. How exactly were you going to run the tests on Android devices, assuming the `kernels:all` build would succeed with this change?

**Comment by freedomtan** [comment]
We are developing a delegate. Currently what we do is

1.  build all kernels
```
$ bazel build --config android_arm64 //tensorflow/lite/kernels:all
```

2.  push all libraries to device
```
$ adb shell mkdir /data/local/tmp/test_libs
$ adb push bazel-bin/_solib_arm64-v8a/lib*.so /data/local/tmp/test_libs
$ adb push bazel-bin/_solib___Caarch64-linux-android-clang8.0.7-libcpp/lib*.so /data/local/tmp/test_libs
```

3. push what we want to test to device, e.g., 
```
$ adb push bazel-bin/tensorflow/lite/kernels/*_test /data/local/tmp/kernel_tests
```
4.  then we can run test_add, 
```
$ adb shell LD_LIBRARY_PATH=/data/local/tmp/test_libs /data/local/tmp/kernel_tests/add_test
```

5. we actually do something like the following on devices
```
$ cd /data/local/tmp
$ for a in kernel_tests/*_test; do LD_LIBRARY_PATH=test_libs $a --use_our_delegate=1 >> kernel_test_our_delegate_results.txt 2>1&; done
$ for a in kernel_tests/*_test; do LD_LIBRARY_PATH=test_libs $a --use_nnapi=1 >> kernel_test_nnapi_results.txt 2>1&; done
$ for a in kernel_tests/*_test; do LD_LIBRARY_PATH=test_libs $a  >> kernel_test_tflite_results.txt 2>1&; done
```
6. pull those logs to a linux or mac and comparing them

File: tensorflow/lite/kernels/BUILD [hunk 1]
```diff
@@ -150,6 +151,12 @@ cc_library(
         "@com_google_absl//absl/types:optional",
         "@com_googlesource_code_re2//:re2",
     ],
+    linkopts = tflite_linkopts() + select({
```
Comment (terryheo) [inline tensorflow/lite/kernels/BUILD:154]: how about
```
    linkopts = select({
        "//third_party/tensorflow:windows": [],
        "//conditions:default": ["-lm"],
    }),

```
?

File: tensorflow/lite/kernels/kernel_util_test.cc [hunk 2]
```diff
@@ -14,11 +14,11 @@ limitations under the License.
 ==============================================================================*/
 #include "tensorflow/lite/kernels/kernel_util.h"
 
-#include <math.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 
+#include <cmath>
```
Comment (terryheo) [inline tensorflow/lite/kernels/kernel_util_test.cc:21]: why is this needed?

**Comment by yyoon** [comment]
What about building them with `--linkopt=-lm` from the command line?

    bazel build --config=android_arm64 --linkopt=-lm //tensorflow/lite/kernels:all

Would this resolve your issue?
Again, I'm a bit hesitant because of the potential unintended side effects this change might cause to any downstream binaries.

**Comment by terryheo** [comment]
@yyoon using "--linkopt=-lm" looks strange to me since user need to know library dependencies. I think this kind of changes are fine and we already have the similar code. 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/BUILD#L808
Also all changes are about fixing test targets (acceleration_test_util_internal, eigen_support_test, kernel_util_test)

**Comment by yyoon** [comment]
Ah, I missed that we already have similar conditions in other places. Sounds good to me.

File: tensorflow/lite/kernels/BUILD (hunk 1 above)
Comment (freedomtan) [reply tensorflow/lite/kernels/BUILD:154]: OK, will do.

File: tensorflow/lite/kernels/kernel_util_test.cc (hunk 2 above)
Comment (freedomtan) [reply tensorflow/lite/kernels/kernel_util_test.cc:21]: that's an interesting / weird case, when building with certain NDK, I saw some non-ANSI/ISO libm function not declared in `<math.h>` but in `<cmath>. I'll check if I can find the case. Otherwise, I'll revert it.

Comment (freedomtan) [reply tensorflow/lite/kernels/kernel_util_test.cc:21]: it was a `std:ldexp()` problem on when building with NDK r20 on Linux at that time. It's no longer a problem in recent master branch.

**Review by terryheo — DISMISSED** [review]
LGTM

**Comment by gbaned** [comment]
@freedomtan  Can you please address Ubuntu Sanity errors? Thanks!

**Comment by terryheo** [comment]
FYI, this is the error log.

```
=== Sanity check step 4 of 15: do_buildifier (buildifier check) ===

Running do_buildifier on 527 files

tensorflow/lite/kernels/BUILD # reformat callsort

buildifier took 0 s

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
tensorflow/lite/kernels/BUILD:
149a150,153
>     linkopts = select({
>         "//tensorflow:windows": [],
>         "//conditions:default": ["-lm"],
>     }),
155,158d158
<     linkopts = select({
<         "//tensorflow:windows": [],
<         "//conditions:default": ["-lm"],
<     }),
279a280,283
>     linkopts = select({
>         "//tensorflow:windows": [],
>         "//conditions:default": ["-lm"],
>     }),
287,290d290
<     linkopts = select({
<         "//tensorflow:windows": [],
<         "//conditions:default": ["-lm"],
<     }),
465a466,469
>     linkopts = select({
>         "//tensorflow:windows": [],
>         "//conditions:default": ["-lm"],
>     }),
472,475d475
<     linkopts = select({
<         "//tensorflow:windows": [],
<         "//conditions:default": ["-lm"],
<     }),
exit status 1
Please fix manually or run buildifier <file> to auto-fix.
```

**Comment by freedomtan** [comment]
@terryheo  Thanks. Done. Sorry for forgetting to run `buildifier` before git push.

**Review by terryheo — DISMISSED** [review]
LGTM

File: tensorflow/lite/kernels/BUILD [hunk 3]
```diff
@@ -1,4 +1,5 @@
 load("//tensorflow/lite:build_def.bzl", "tflite_copts")
+load("//tensorflow/lite:build_def.bzl", "tflite_linkopts")
```
Comment (terryheo) [inline tensorflow/lite/kernels/BUILD:2]: could you merge this line with the pervious one?

Comment (freedomtan) [reply tensorflow/lite/kernels/BUILD:2]: done. thanks.

**Review by terryheo — DISMISSED** [review]
LGTM

File: tensorflow/lite/kernels/BUILD [hunk 4]
```diff
@@ -1,5 +1,4 @@
-load("//tensorflow/lite:build_def.bzl", "tflite_copts")
-load("//tensorflow/lite:build_def.bzl", "tflite_linkopts")
+load("//tensorflow/lite:build_def.bzl", "tflite_copts", "tflite_linkopts")
```
Comment (terryheo) [inline tensorflow/lite/kernels/BUILD:1]: oops, I've just realized that we don't use tflite_linkopts() now. could you revert the change?

**Review by terryheo — APPROVED** [review]
LGTM
